DM-8723: Task template

mssgill opened the pull request:
For comments on the more filled in Task Template.

Commit:
Make Task template consistent with example tasks

This is incorporating our experience with the sfp tasks and bringing
it into the Task template.

Commit:
Add needed files to standalone template dir build

All the ones we were using in the sfp subdir, with small modifications
as needed.

Commit:
Remove See Also header

Commit:
Rename template.forTasks to ExampleTask

At JSick's suggestion, makes it better in more parallel.

Commit:
Modify ExampleTask entry in index.rst

To be consistent with name change inside Task Template file.

Commit:
Modify ExampleTask to Task Template in index.rst

Wanted this to be the name instead.

Commit:
Remove the 'Entrypoint' section

We found this is not being useful for us, everything in the `run`
interface is now incorporated into the API Usage page which is linked
to from the SeeAlso box above.

Commit:
Update Task Template with sfp Task learnings

This is the first full draft of what the Task Template will look like
as handed off to DM developers.  It contains many modifications and
much more content from we have learned while writing out the 4 sfp
Tasks in detail by hand.  It's written largely in the imperative form
as seemed appropriate when passing on, and further delineates the
sections that will be filled in automatically, and those that will
need docstrings actually written (which is the majority of it).  There
are a number of comments in the rst in this version for places that
there is not yet full clarity on what's to be done.

jonathansick reviewed (commented):
There's a few layers of things that can be addressed here:

1. The overall format for this template documentation.
2. The specific documentation/guidance presented here.

This comment is mostly about (1), but we'll get to (2) later.

I know we all talked about dropping template instructions into the template itself, but after seeing it implemented, I'm not sure it's a sustainable approach.

It's kind of hard to know, at a glance, what content is part of the template itself, what part of the template needs to be modified, and what content is instructions that need to be deleted by the author.

Here's my proposal:

- The template ought to be stripped down to the section headings and directives that an author will work from. Inside each section there's a reStructuredText **comment** that (1) names the section, (2) in a sentence says what the section is for, and (3) links to the section in the Developer Guide that documents and gives guidance for how to write that section.

- Put the instructions in the Developer Guide. We'll create a new directory called `docs/templates` and the page where this template is documented will be called `docs/templates/task.rst`. In this `task.rst` page, each component of the task topic will have a header. This header has a permalink that the template itself can use. Then in each component's section we provide (1) guidance, and (2) example reStructuredText. Doing this, the template documentation will look a lot like https://dmtn-030.lsst.io/#task-topic-type In fact, once the template documentation is in place, we can deprecate those sections of DMTM-030.

mssgill commented:
@jonathansick -

Ok, you are right that the idea of what's here already in the template and what the developer needs to actually write is blended together, in the way i've shown it.   So for now i will work on demarcating the two parts very clearly in the Task Template itself, but then i do think your idea of moving later to primarily including in the Template itself only the 'boilerplate' structure, then links to the specific instructions for devs in the Dev Guide is a good division and way to present the info.

Commit:
Separate descriptive parts and 'parts to fill'

Did this using the 'note' directive for now, and it does already look
way better and more accessible for what a developer would need to see.
Need to figure out if this is the directive we want to stay with, and
how the language should change also.

Commit:
Change note to ToDo directive

Seems like the more proper one to use in the context of asking the
devs to fill in their sections.  Also bolded section titles to more
clearly indicate the dif parts.

mssgill commented:
@jonathansick - I've now used the ToDo directive to separate off the parts where we want the devs to fill in the docstrings, and i also bolded section titles to more clearly indicate the dif parts so that the template does look like it has clearer structure now and separation between the 2 parts but.. I agree that even to my eye it looks messy and jumbled together a bit when done like this.  I will next work on being much more exact about what sections the devs would be filling in their docstrings, and with what content, before attempting to separate out into two parts as you suggest.

Commit:
Detail locations for docstring insertion

For each section, after looking closely at where the doxygen
docstrings are written, i entered analogous (first draft) directional
info for the new docs 'regime'.

Commit:
Split the task template into 2 pieces

Following the convo with JSick (in Slack and in the PR here), this is going to separate the pieces that show how the page is to be structured (the struc_template) and the instructional one that gives specific guidance on what parts the devs will be filling out (the instruc_template).

Commit:
Insert :orphan: true to suppress warnings

Standard one, to suppress compile-time warnings.

Commit:
Strip out instructions to devs

Reverting to the form before, closer to DMTN-030, just the structural
outline now and will insert instructions in other instruc code.

Commit:
Separate structure and guidance pages for Tasks

This is the first draft of a separated version now of the docs where
i've stripped the structure template down, and put links into the
appropriate header locations for the guidance doc page (which will be
the one to go into the dev guide).

Commit:
Make pointers in Structure Template rst comments

This is so this template could now be copied and dropped straight into
some dir, and then filled in there by the dev (deleting all the rst
comments).

Also changed wording, including about docstrings.

Commit:
Clarify content for strings vs. docstrings

This is primarily in the Guidance doc, where comparing with the
doxygen placement of docstrings, i now refer more precisely to what
goes in which part of the strings of the code -- which will be picked
up automatically anyway, so devs won't need to do anything 'extra'
here, and that should be stated.  But at least, it's clear now what
should be said about the string parts in the guidance doc.

Commit:
Move comments that sections are autofilled earlier

Want to make very clear to devs what they need to do and not do, so we
move the comments on each section to the start, so they know what to
concentrate on.

Commit:
Remove reference to 4 sfp Tasks

We can't assume the current protos will stay in the form they are
currently, so we need the Task Template to be standalone for now.

We'll next bring in examples though from each of them, into the
Guidance Doc.

Commit:
Insert examples for each sec of Guidelines Doc

Used mostly procCcd, and this is the first draft form.

Commit:
Switch all secs to imperative

Want to really instruct devs on *what* to fill in on that section, and
we say that at the outset now.

Also, removed extraneous comments, and changed description of parts
that will be auto-filled from what will need dev input.

Commit:
Remove extraneous comments from Guidance Doc

There were many rst comments that weren't relevant for devs, but just
as we were deciding content for this page, have removed them now.

Commit:
Lengthen procCcd abbreviation

Put in full name for it everywhere it appears.

Commit:
Insert Latex example explicitly

Want devs to see how this example text actually looks, so give it in full detail.

Commit:
Wrap lines at 80 chars

This is simply to make it easier to read the rst instructions for devs
in editors (specifically suggested by @jmeyers).

Commit:
Change Latex line to be inside raw html directive

This is the way it renders best in the page.

Commit:
Bring in explicit IsrTask example guidance

Use it as an example and bring in code blocks that we can use in the
Guidance doc without getting unwieldy.

Also, again, word wrap all lines at 80 char.

jonathansick commented on templates/instruc_template.forTasks.rst:
```diff
@@ -0,0 +1,236 @@
+.. Based on: https://dmtn-030.lsst.io/v/DM-7096/index.html#task-topic-type, with learnings from the 4 sfp pages built in branch DM-8717
+
+########################
+Guidance for ExampleTask
```
I recommend a more useful title:

```
Writing Task Topics
```

I both prefer this grammar, and also, `ExampleTask` has no real meaning.

mssgill replied:
Ok (we had moved to `ExampleTask` earlier for this header, and initially this document was meant to be arrived at *from* the `ExampleTask` Template itself, but if this is the first page they are coming to, as i can see you would intend for them, then i see what you mean about how it's less descriptive than your suggestion)

jonathansick commented on templates/instruc_template.forTasks.rst:
```diff
@@ -0,0 +1,236 @@
+Guidance for ExampleTask
+########################
+
+We give guidance here for developers on what to write into the
```
I'd like to see some development on the language here.

> We give guidance here for developers on what to write into the [..]

I think phrases like this can be iterated upon to be friendlier on the ear.

mssgill replied:
Guess it's really covered by the way you rephrased it above, so no reason to load more words in when the essentials have been stated once.

jonathansick commented on templates/instruc_template.forTasks.rst:
```diff
@@ -0,0 +1,236 @@
+covered in `this section of the Developer Guide
+<https://developer.lsst.io/docs/py_docs.html>`_ .
+
+Below, the normal text lists notes on the each of the sections of the
```
What's "normal text?"

mssgill replied:
I'd meant as opposed to in the set off 'Notes' boxes..

jonathansick commented on templates/instruc_template.forTasks.rst:
```diff
@@ -0,0 +1,236 @@
+direction to the developer of what to include into the reST for that
+section.
+
+To be explicit, we will give some examples for what to put into the
```
I don't think it's necessary to point out at this location where the examples come from.

mssgill replied:
Ok.. had been thinking that it could be of value if they did have the full sfp examples to refer to at some point, but i guess you're right that it's not needed at this point.

jonathansick commented on templates/instruc_template.forTasks.rst:
```diff
@@ -0,0 +1,236 @@
+To be explicit, we will give some examples for what to put into the
+sections taken primarily from `ProcessCcd` specifically, but from other tasks as well.
+
+[NB: The structure is substantially different from how docstrings were
```
A couple points:

- Let's avoid "NB" and other Latin. We're not writing a stuffy academic paper here.
- I don't think we need to point to previous standards here.

mssgill replied:
Ok - i find NB pretty standard and shorthand usage, but in any case we don't really need to say anything like this at all, here.  I'd wondered if it might be seen to be useful, thought it could be because many of the devs will be familiar with how the doxygen pages were written and we can let them know now that the new 'regime' will be different but fine to keep it leaner and leave this out.

jonathansick commented on templates/instruc_template.forTasks.rst:
```diff
@@ -0,0 +1,236 @@
+populated for DM's previous Doxygen documentation, for those familiar
+with it.]
+
+.. _intro:
```
Remember to namespace your anchor link since there form a global space, which is especially crowded in the Developer Guide. Instead, try:

```rst
.. _task-topics-intro:
```

mssgill replied:
Ah, yes, i'd been operating more in standalone thinking, but that's completely right for when this gets imported into the DevGuide.  Changed all of these.

jonathansick commented on templates/instruc_template.forTasks.rst:
```diff
@@ -0,0 +1,236 @@
+.. _intro:
+
+Introduction
+=============
```
Be careful with the header underlines. This underline is one character too long.

mssgill replied:
Right, altered, checked all of the rest so they're ok.

jonathansick reviewed (commented):
Here's some initial feedback on what you've got so far. I'll give more feedback on instructions for each topic component next.

## Feedback on file naming conventions:

Don't use periods in file names. Also avoid weird mis-spellings and a unusual abbreviations. For example,

```
`templates/instruc_template.forTasks.rst`
```

should be named:

```
templates/writing-task-topics.rst
```

The reason this is important is because the rST file name becomes a URL, so it's good to have all lower case, hyphen-separated names that are also clean and descriptive.

(I know I've used underscore separators in existing docs but I'm intending to migrate those URLs to use hyphens too).

## Feedback on the Intro

I think the intro is burying the lead regarding how to use the template, including where to put the template in a package's documentation. Try this:

> [H1] Writing task topics
> This page describes how to document science pipelines tasks for the pipelines.lsst.io documentation project.
>
> [H2] Using the template
>
> We provide a standardized template for all task documentation pages. This required template ensures consistency, and also gives you tools for automatically extracting documentation from your code.
>
> You can get the template from here [insert link].
>
> [H3] Installing the template
>
> To install the template, copy it into the `doc/tasks` directory of your stack package. Then name the file after the task class, with a ``.rst`` extension. For example: ``doc/tasks/ProcessCcdTask.rst``.
>
> [H3] Filling out the template
>
> The template includes brief instructions as reStructuredText comments. Each component of the documentation page is described in more detail below, on this page.

mssgill commented:
@jonathansick - got ya, I'm in concordance on most all your suggestions, and appreciate that you explained the reasoning behind some of them too which would not have been obvious me at first glance (e.g. periods in filenames, extended anchor names etc.)  

Have now made changes according to these suggestions, and pushed them here.

Commit:
Rename template doc files

To make consistent with others, and to make it easier later when names
are made into URLs.

Commit:
Make intro more directly useable

Mostly strip it down and break into headings so it's easier to
visually parse at a quick reading, and change some wording
(implementing some of the suggested changes from the PR now).

Commit:
Change names of anchors

This is done for compatibility with the global namespace when this
gets imported into the DevGuide.  Changed anchor names in the Guidance
doc and correspondingly links in the Template.

Commit:
Put in correct filename for Task Template

Needed to do this since we changed the filename.

Commit:
Remove extraneous rst comments

Didn't need to keep these comments since we've altred the core intro
text now.

Commit:
Remove empty line warnings at build

This was for a list that had no entries, and though we aren't actively
displaying these files at the moment, better to suppress the warnings
at build time still.

Commit:
Replace 2 other outdated links

Had missed these two to the old 'instruc' name of the Guidance doc, replaced now.

Commit:
Insert lines to ignore scipy for intersphinx

Because of the long latencies, but this should be only a temporary
needed fix until it's fixed on their end.

See: https://community.lsst.org/t/how-docs-scipy-org-and-intersphinx-affect-your-doc-builds/1717

jonathansick commented on templates/index.rst:
```diff
@@ -0,0 +1,27 @@
+
+.. toctree::
+
+  Task Template <template-for-tasks>
```
We won't be rendering templates in Sphinx. They may even end up the the https://github.com/lsst/templates repo — they're just files that we let a developer download and then fill in. What we *are* rendering is the documentation for the template, and that's what will end up in developer.lsst.io.

mssgill replied:
Right, i did get this from your last comments, but i now cleaned up, by only linking to the Guidance Doc for the template in the toctree (in index.rst).

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+   https://dmtn-030.lsst.io/v/DM-7096/index.html#task-topic-type, with
+   learnings from the 4 sfp pages built in branch DM-8717
+
+Please fill in the below sections with the contents as described in
```
This should be a rST comment. Again, nothing should be rendered unless it should be rendered in the deployed document.

mssgill replied:
Ok, changed.

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+:orphan: true
```
This shouldn't be here since it won't be used in production documents.

mssgill replied:
Right, i had inserted it temporarily to suppress the warnings that the file isn't included in any toctree when we build in this dir, until we move it to the right place, but i removed the line, and we'll just be aware of this being a temporary issue.

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+
+
+########################
+Structure of ExampleTask 
```
There should be guidance for how to structure a title, since the title won't just be "Structure of ExampleTask," but we certainly do want consistently structured titles. A way to do this is:

```rst
.. The title of the page is the name of the Task class (ProcessCcdTask, for example).
.. Also, update the section anchor to have the same name.
.. See https://developer.lsst.io/writing/user-guides/task-topics.html#title for details.

.. _TaskClassName:

#############
TaskClassName
#############
```

This way of documenting a template **component** does two things:

1. Provides inline documentation on how the template should be modified/filled in. But there's no extraneous real content. Just the title and anchor that the developer needs to modify.
2. Provides a link to the template's documentation in the Developer Guide (of course, the URL itself is TBD).

This same pattern should be used for the rest rest of the components in the template. I think I mentioned this before in either our meetings or in another review.

mssgill replied:
Yes, but when i started this template i was mixing the 2 pages (template + guidance) and this is residue from that (sorry).  Corrected with above now.

About the URL: there is useful info at the "Title" section of https://dmtn-030.lsst.io/#task-topic-type (whose internal link needs to be corrected btw, right now the first 3 internal links:  

Title
Summary sentence
Processing sketch

jump to sections later than the actual ones), but we won't be linking to that directly, right, so i guess we'll want to integrate that with the relevant parts of the Guidance doc and insert it into the TBD location (?)

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+
+..  b) A method within a task.
+
+`Guidance for the Introduction Section  <writing-task-topics.html#task-topics-intro>`_ .
```
The URL should also be in a comment, not as reStructuredText.

mssgill replied:
Ok, i'll do that on all these (i went back and forth a bit when i was actually rendering the page and found it useful to click on sections myself to go to them), that's consistent with the idea that it will never be actually viewed as html by devs.

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+`Guidance for the Introduction Section  <writing-task-topics.html#task-topics-intro>`_ .
+
+
+.. - Insert Module Membership here:
```
Since the module membership component has such a consistent structure, you can include templated content, with a comment above.

```rst
.. This component mentions and links to the task's parent module.
.. Modify the module-anchor in the ref to point to the module page.
.. See <URL> for details.

This task is implemented in the :ref:`module-anchor` module.
```

jonathansick replied:
Also, I'm in favor of having the documentation comment precede the template content as a way of introducing and demarcating each component. The comment should also be *compact*, like in my examples so it's easy for a developer to delete them as the section is filled in.

mssgill replied:
Ok, I inserted the above (except instead of inserting 'See <URL> for details' leaving the link in currently to that section in the Guidance doc, but modifying wording there some), and removed extraneous blank lines everywhere (easier to *read* with blank lines in i think, but more of a pain to go scroll through and delete them for devs later, of course.)

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+
+`Guidance for the Module Membership Section  <writing-task-topics.html#task-topics-module>`_ .
+
+.. Insert material for the SeeAlso Box here:
```
I'd insert the `.. seealso::` directive here, but precede it with a documentation comment.

mssgill replied:
Ok, i inserted the rst comment "Insert material for the SeeAlso Box here " after the directive, otherwise just compacted it down.

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+Configuration
+=============
+
+.. - This section will be autofilled -- it describes the task’s
```
You can leave the Configuration header, but there isn't any need for the two subheads nor the comments. A developer won't need to modify the *contents* of this section at all.

If anything you can put a comment *preceding* this section telling developers to modify the section's anchor to have the format

```rst
.. TaskClassName-config:
```

(and include that template content in the template itself).

mssgill replied:
Ok, stripped this down by removing the Retarg/Params subsections.

Also put in this anchor, and inserted a comment before it.

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+
+`Details about the Parameters Subsection  <writing-task-topics.html#task-topics-params>`_ .
+
+Python usage
```
Also put a template for the section anchor here.

```rst
.. _TaskClassName-python-usage:
```

mssgill replied:
Done, though you didn't mention deleting the (extraneous) comments for all these other sections that will be autofilled here and below (other than for Debugging), so i haven't done that yet (though i compacted them all).

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+Python usage
+============
+
+Class initialization
```
Also put a section anchor here, formatted similarly.

```rst
.. _TaskClassName-class-init:
```

mssgill replied:
Done

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+
+`Details about the Class initialization Subsection  <writing-task-topics.html#task-topics-initzn>`_ .
+
+Run method
```
```rst
.. _TaskClassName-run:
```

mssgill replied:
Done

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+`Details about the Run Method Subsection  <writing-task-topics.html#task-topics-run>`_ .
+
+
+Debugging
```
```rst
.. _TaskClassName-debugging:
```

mssgill replied:
Done

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+Debugging
+=========
+
+.. This section will be autofilled also -- the content is filled in
```
Again, all of this content can be deleted. Once the directive is available there will be instructions on how to configure the directive, but that's it.

mssgill replied:
Stripped out

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+
+`Details about the Debugging Section  <writing-task-topics.html#task-topics-debug>`_ .
+    
+Examples
```
```rst
.. _TaskClassName-examples:
```

mssgill replied:
Done

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+
+.. - Fill in a self-contained example of using this task that can be tested by any reader.
+
+`Guidance for the Examples Subsection  <writing-task-topics.html#task-topics-examples>`_ .
```
Again, the URL should be in a comment, not in reStructuredText. I also recommend that the instructional comment precede the component, as in my earlier examples.

mssgill replied:
Yep, commented out all URLs.

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+`Guidance for the Examples Subsection  <writing-task-topics.html#task-topics-examples>`_ .
+   
+Algorithm details
+====================
```
Check the length of your underlines.

mssgill replied:
Yeah, sorry - i try to watch for these, but some slip through. 

(However, i used to get warnings when i did the build whenever this happened, but no longer do..  my conf.py is pretty minimal so i don't think there's anything in it which is suppressing the warnings, but something else must've changed, and it appears to be the version of Sphinx, as you mentioned when we spoke last week.)

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+
+`Guidance for the Examples Subsection  <writing-task-topics.html#task-topics-examples>`_ .
+   
+Algorithm details
```
```rst
.. _TaskClassName-algorithm:
```

mssgill replied:
Done

jonathansick commented on templates/template-for-tasks.rst:
```diff
@@ -0,0 +1,171 @@
+   the math can be nicely displayed and written in straight Latex
+   (through the **math** directive of reST).
+
+`Guidance for the Algorithm Details Section  <writing-task-topics.html#task-topics-algorithm>`_ .
```
Again, the URL should be in a reST comment, and precede the component.

mssgill replied:
Commented.

jonathansick commented on templates/template.forCLT.rst:
```diff
@@ -1,11 +1,8 @@
-
+:orphan: true
```
I recommend that you not add/commit additional stubs like this one unless your ticket is dedicated to completing that work.

mssgill replied:
Ok, these were leftover from when we were going to work on all the templates at once (iirc), removed now, along with the others.

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+Writing Task Topics
+###################
+
+This page describes how to document science pipelines tasks for the
```
I'm about to formally RFC this, but at this stage I recommend (especially for documentation specialists) that you use one-sentence-per-line formatting for your reStructuredText prose content (rather than hard-wrapping at an arbitrary line length). That's the only way to effectively manage text content in Git and in GitHub PRs.

mssgill replied:
Have done this (made more sense as a practice once i realized this is so the git history shows the exact line that's been modified, vs. the whole paragraph.)

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+###################
+
+This page describes how to document science pipelines tasks for the
+`pipelines.lsst.io <http://pipelines.lsst.io>`_ documentation project.
```
https

mssgill replied:
That's curious - i assume i would've cut and pasted this in, and it defaults to the secure URL (in fact, if you either click on it, or try to insert the http into the URL window, they will both revert back to the https site), so not sure how it crept in, but i corrected it, anyway.

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+
+The template includes brief instructions as reStructuredText
+comments. Each component of the documentation page is described in
+more detail on this page, below,.
```
> below,.

mssgill replied:
Corrected

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+comments. Each component of the documentation page is described in
+more detail on this page, below,.
+
+
```
Don't forget documentation on the title component.

mssgill replied:
Inserted this (kind of duplicates in this case the info right in the Template itself).

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+
+.. _task-topics-intro:
+
+Introduction
```
Looking at this, I recommend we explode the "Introduction" component into these discrete components:

1. Summary/context (1-2 sentences).
2. Summary of logic and algorithm.
3. Module membership.
4. See also

In this document each component should have its own section (an H2, with `=` underlines).

More generally, the documentation for each template component should follow this format:

1. Section title (with preceding anchor so that URLs in the template can directly link to individual components).
2. Description and instructions.
3. Example, as appropriate, using `.. code-block:: rst` to show the example source.

mssgill replied:
Ok... this is a more substantial change, expanding a number of the subsections in the intro, but i realize that it was getting kind of top-heavy so that it makes sense.

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+
+- **Summary/context (1-2 sentences):**
+
+.. Note:: Write this section concisely, as it should be very brief,
```
This content shouldn't be in a note—make it a plain paragraph.

Secondarily, the formatting of your notes is a bit odd.

```rst
.. note::

   Note content.
```

is more canonical.

mssgill replied:
Ok, made it straight text, and after our talk where you mentioned that studies show that all 'colored boxes' are ignored by readers generally, i decided to remove all 'Note' boxes in fact.

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+
+.. code-block:: rst
+
+ ProcessCcdTask is a command line task which executes the processing
```
The content of your `code-block` directive, along with **all** directives [must be indented three spaces](https://developer.lsst.io/docs/rst_styleguide.html#indentation).

mssgill replied:
Ok, changing all.

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+
+.. _task-topics-config:	  
+    
+Configuration
```
For components like this that are entirely rsT directives the content should just state two things:

1. How to format the anchor link.
2. How to configure the directive.

Anything else is extraneous and you can likely leave the content for (2) as "TBD" for now.

mssgill replied:
Ok.. isn't (1) already explained within the task-template?  And should we really write TBD explicitly..?  

Have otherwise just deleted all content in these sections, since it's not directing devs to do anything, as you say.

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+Python usage
+============
+
+.. _task-topics-initzn:
```
**I strongly discourage unusual abbreviations/contractions.** `initzn` is unusual. Use `init` instead.

mssgill replied:
Yeah, sorry - i by default use abbreviations that are very familiar to me in links etc., forgetting at times this is going to be embedded in larger structures and every word will be open to public scrutiny.  Will try to be more aware of this in the future.

(Trying to be mostly canonical in my writings here on Github, too, mostly, except that i am so used to using lower-case "I" that i stay with that usage, for now.)

Changed.

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+--------------------
+
+- The interface declaring the instance of the class will be
+  picked up automatically from the interface of the `__init__` method
```
Careful about your inline roles here. In this case you need a code literal as you're not referring to a specific `__init__` API object:

```rst
``__init__``
```

mssgill replied:
Ah, right -- but from your suggestion above, i am just stripping everything that isn't directions to devs at this point, and not leaving any of the previous content of these auto-filled sections, but am putting in directions for how to insert documentation for the variables in these sections.

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+  
+.. _task-topics-run:
+	  
+Run method
```
Again, same as https://github.com/lsst-dm/scipi_docs_prototype/pull/5/files#r106199283

mssgill replied:
Stripped out

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+  
+.. _task-topics-debug:
+
+Debugging
```
Same as https://github.com/lsst-dm/scipi_docs_prototype/pull/5/files#r106199283

mssgill replied:
Stripped out

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+
+.. code-block:: rst
+
+  .. code-block:: python
```
I strongly recommend you use the canonical three space intent to align your directive content.

mssgill replied:
Ok - i tried to be careful in all cases (this one is a double indent and requires 6 for the python code lines), and will watch this henceforth.

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+
+.. code-block:: python
+		
+       output = isrTask.run(rawExposure, dark=darkExposure, flat=flatExposure)
```
Again, align your directive content.

mssgill replied:
Corrected

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+.. _task-topics-algorithm:
+   
+Algorithm details
+====================
```
Watch for title underline lengths.

mssgill replied:
Changed (with same comment as above on this)

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+
+This should be written in detailed form and can refer to prior written
+documentation as long as it is accessible by all potential code users.
+Mathematical notation can be used here and written in LaTex through the `math`
```
> Latex

jonathansick replied:
Again, be careful about your inline text roles. ``math`` isn't a Python API object.

Probably a better approach is:

```rst
through the :rst:directive:`math directive <sphinx:math>`.
```

And add Sphinx to the `intersphinx` configuration in `conf.py`.

The mention to the Developer Guide is good, though.

mssgill replied:
The official spelling appears to be LaTeX actually (https://en.wikipedia.org/wiki/LaTeX) but i'll simplify and go with Latex.

Accepted your changes for referring to 'math' directive.

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+Examples
+========
+
+.. Note:: Making this can be a substantial job which requires writing
```
This content doesn't need to be in a `note` directive and needs to be re-written slightly to be more purposeful. I think the main thing you want to get across is *what the example is for*.  We want to establish that an example should be self contained, allowing a user to follow some steps and have the task work. Any data and configuration should be included in the example.

mssgill replied:
Ok, removed, and reworded.

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+`exampleIsrTask.py`, then describe it with reST.
+
+In this particular case, we need to use some functions which are
+normally in the `utils.py` class, and to make it more transparent, we
```
What `utils.py`? If it's important enough to talk about, you need to fully specify what it is.

mssgill replied:
Changed wording -- this is actually Simon's purpose-written exampleUtils.py for this example (and a few others)

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+
+Next, we describe some of the details for the content of `exampleIsrTask.py`.
+
+Where needed, when describing code, the python code block directive can be used as so:
```
Also mention task configuration, in addition to code in general, since task configuration is a big part of what these examples will show.

mssgill replied:
Ok, stated this explicitly

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+
+.. code-block:: rst
+		
+ The first line: `isrConfig = IsrTask.ConfigClass()` indicates this is
```
You want a code literal here:

```rst
``isrConfig = IsrTask.ConfigClass()``
```

mssgill replied:
Right, changed

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+This should be written in detailed form and can refer to prior written
+documentation as long as it is accessible by all potential code users.
+Mathematical notation can be used here and written in LaTex through the `math`
+directive of reST, for details see `this section of the Developer Guide that
```
Probably good form to shorten the link text:

```rst
[..] for details see `the reStructuredText Style Guide <https://developer.lsst.io/docs/rst_styleguide.html#rst-math>`__.
```

and it's essentially assumed you're linking to the appropriate section.

mssgill replied:
Ok.. though you're sure you'd rather i refer to that section as the "rST Guide" vs. "the rST section of the Dev Guide"?

jonathansick commented on templates/writing-task-topics.rst:
```diff
@@ -0,0 +1,240 @@
+describes how to insert mathematical expressions
+<https://developer.lsst.io/docs/rst_styleguide.html#rst-math>`_ .
+
+Here is an example of the syntax for how one would insert an equation (from IsrTask):
```
> for how one would

This should set off alarms for writing that could be tightened up.

mssgill replied:
Ok, compacted it.

jonathansick reviewed (commented):
Here's my latest round of comments on the task template and associated documentation.

The inline comments should give you a clear sense of what I'm looking for from this work but the common themes are:

1. Reduce the template to just a template. It's not something that will ever be rendered on its own.
2. All instructions embedded in the template should be comments preceding the component, ending with a URL the components section of the task template's documentation.
3. In the task template documentation, each template component should have its own section. The "Introduction" in particular needs to be split up into individual template components.
4. Documentation for directive-based template components can be reduced to (1) instruction on the format of the section anchor link and (2) a TBD placeholder for the directive's configuration.

Commit:
Remove all extraneous files

Didn't need any of the extra templates in this dir currently as we
hone in on the Task Template, so removed them even in commented form
from the index.rst file.

Commit:
Respond to PR #5 - compacting, cleaning

There are multiple changes in this file, detailed at the PR page on
github:

 https://github.com/lsst-dm/scipi_docs_prototype/pull/5#pullrequestreview-27091993

Commit:
Remove other template stubs from repo

Didn't need any of the extra templates in this dir currently as we
hone in on the Task Template, so have removed them for now.

Commit:
Explode the Intro section

This is primarily to expand into subsections, as per suggestion in PR

Commit:
Implement changes in all subsections

Number of changes throughout, see PR #5 (3/15/2017 comments and
responses) for details.

Commit:
Make into a text vs. rst doc

Took out actual section headers that will disply, put anchors at the
end of the doc lines, compacted each section

Commit:
Clean up template links

Add in links to Guidance Doc.

Commit:
Insert example snippets in multiple sections

Want to make easier and more useable for devs.

Commit:
Insert full absolute URL to Guidance Doc locations

This will be temporary until it's ascertained where we actually want
to point to, and that comment has also been inserted.

Commit:
Indent properly rst code blocks

Had accidentally left this out so block was not being rendered properly.

Commit:
Insert comments into auto-populated sections

General currently, but we will next describe in more detail what rst
to include into strings that describe variables, flags etc. that will
get pulled into these sections when the full doc structure is built.

Commit:
Insert content in all auto-populated sections

Used primarily current content, need to convert to rst format though.

Commit:
Enter lines to suppress warnings during build

Specifically for the seealso box being empty and the template file not
going into the toctree.

Commit:
Emphasize that procCcd is just an example

Put in a comment that we don't mean for procCcd to be the paradigmatic
code, it's just one example we picked and devs should be able to adapt
their docs appropriately from the example.

Commit:
Old commit

Commit:
Add debug file that works

Commit:
Add altered  exampleStatsTask file

Commit:
Remove files that should be in personal space
